### PR TITLE
XML serializer's type initialization is vulnerable to concurrency

### PR DESCRIPTION
**Pull request by Scooletz**
This PR backports XmlSerializer #5166 fix to version 5.

I did the backporting manually, please review it carefully.

### CODE REVIEW

File: src/NServiceBus.Core/NServiceBus.Core.csproj [hunk 1]
```diff
@@ -76,7 +76,9 @@
     <Reference Include="System.Web.Services" />
     <Reference Include="System.Windows.Forms" />
     <Reference Include="System.Xml" />
-    <Reference Include="System.Xml.Linq" />
+    <Reference Include="System.Xml.Linq">
+      <HintPath>..\..\..\..\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\System.Xml.Linq.dll</HintPath>
```
Comment (timbussmann) [inline src/NServiceBus.Core/NServiceBus.Core.csproj]: 🔥 ?

File: src/NServiceBus.Core/Serializers/XML/XmlSerializerCache.cs [hunk 2]
```diff
@@ -1,20 +1,36 @@
-namespace NServiceBus.Serializers.XML
+namespace NServiceBus
```
Comment (timbussmann) [inline src/NServiceBus.Core/Serializers/XML/XmlSerializerCache.cs]: hm, let's not change the namespaces

File: src/NServiceBus.Core/Serializers/XML/Serializer.cs [hunk 3]
```diff
@@ -372,13 +372,14 @@ void Write(StringBuilder builder, Type t, object obj)
                 return;
             }
 
-            IEnumerable<PropertyInfo> properties;
-            if (!cache.typeToProperties.TryGetValue(t, out properties))
+            Tuple<FieldInfo[], PropertyInfo[]> members;
+            if (!cache.typeMembers.TryGetValue(t, out members))
             {
-                throw new InvalidOperationException(string.Format("Type {0} was not registered in the serializer. Check that it appears in the list of configured assemblies/types to scan.", t.FullName));
+                cache.InitType(t);
```
Comment (timbussmann) [inline src/NServiceBus.Core/Serializers/XML/Serializer.cs]: I think we don't need to backport the runtime initialization logic, so let's keep throwing an exception here?

File: src/NServiceBus.Core/Serializers/XML/Deserializer.cs [hunk 4]
```diff
@@ -274,7 +274,14 @@ object GetObjectOfTypeFromNode(Type t, XmlNode node)
                     type = Type.GetType("System." + n.Name.Substring(0, n.Name.IndexOf(":")), false, true);
                 }
 
-                var prop = GetProperty(t, n.Name);
+                Tuple<FieldInfo[], PropertyInfo[]> typeMembers;
+                if (!cache.typeMembers.TryGetValue(t, out typeMembers))
+                {
+                    cache.InitType(t);
```
Comment (timbussmann) [inline src/NServiceBus.Core/Serializers/XML/Deserializer.cs]: as it looks, the v5 code also didn't use runtime initialization when the type wasn't initialized already?

File: src/NServiceBus.Core/Serializers/XML/XmlSerializerCache.cs (hunk 2 above)
Comment (Scooletz) [reply src/NServiceBus.Core/Serializers/XML/XmlSerializerCache.cs]: I agree.

File: src/NServiceBus.Core/NServiceBus.Core.csproj (hunk 1 above)
Comment (Scooletz) [reply src/NServiceBus.Core/NServiceBus.Core.csproj]: Not initialized R#. will fix

File: src/NServiceBus.Core/Serializers/XML/Serializer.cs (hunk 3 above)
Comment (Scooletz) [reply src/NServiceBus.Core/Serializers/XML/Serializer.cs]: I double checked the fix. You are right. The initialization was introduced in v6 not in the backported fix. Will undo.

**Comment by Scooletz** [comment]
Ping @timbussmann

**Comment by timbussmann** [comment]
the inspection errors need to be ignored.

**Pull request merged**